x/ccv/provider/keeper/partial_set_security_test.go

```diff
@@ -31,6 +37,53 @@ func TestHandleOptIn(t *testing.T) {
 	require.False(t, providerKeeper.IsToBeOptedIn(ctx, "chainID", providerAddr))
 }
 
+func TestHandleOptInWithConsumerKey(t *testing.T) {
+	providerKeeper, ctx, ctrl, mocks := testkeeper.GetProviderKeeperAndCtx(t, testkeeper.NewInMemKeeperParams(t))
+	defer ctrl.Finish()
+
+	// generate a consensus public key for the provider
+	providerConsPubKey := ed25519.GenPrivKeyFromSecret([]byte{1}).PubKey()
+	consAddr := sdk.ConsAddress(providerConsPubKey.Address())
+	providerAddr := types.NewProviderConsAddress(consAddr)
+
+	calls := []*gomock.Call{
+		mocks.MockStakingKeeper.EXPECT().
+			GetValidatorByConsAddr(gomock.Any(), gomock.Any()).
+			DoAndReturn(func(ctx sdk.Context, addr sdk.ConsAddress) (stakingtypes.Validator, bool) {
+				if addr.Equals(providerAddr.Address) {
+					// Given `providerAddr`, `GetValidatorByConsAddr` returns a validator with the
+					// exact same `ConsensusPubkey`
+					pkAny, _ := codectypes.NewAnyWithValue(providerConsPubKey)
+					return stakingtypes.Validator{ConsensusPubkey: pkAny}, true
+				} else {
+					// for any other consensus address, we cannot find a validator
+					return stakingtypes.Validator{}, false
+				}
+			}).Times(2),
+	}
+
+	gomock.InOrder(calls...)
+	providerKeeper.SetProposedConsumerChain(ctx, "chainID", 1)
+
+	// create a sample consumer key to assign to the `providerAddr` validator
+	// on the consumer chain with id `chainID`
+	consumerKey := "{\"@type\":\"/cosmos.crypto.ed25519.PubKey\",\"key\":\"Ui5Gf1+mtWUdH8u3xlmzdKID+F3PK0sfXZ73GZ6q6is=\"}"
+	expectedConsumerPubKey, _ := providerKeeper.ParseConsumerKey(consumerKey)
+
+	err := providerKeeper.HandleOptIn(ctx, "chainID", providerAddr, &consumerKey)
+	require.NoError(t, err)
+
+	// assert that the consumeKey was assigned to `providerAddr` validator on chain with id `chainID`
+	actualConsumerPubKey, found := providerKeeper.GetValidatorConsumerPubKey(ctx, "chainID", providerAddr)
+	require.True(t, found)
+	require.Equal(t, expectedConsumerPubKey, actualConsumerPubKey)
+
+	// assert that the `consumerAddr` to `providerAddr` association was set as well
+	consumerAddr, _ := ccvtypes.TMCryptoPublicKeyToConsAddr(actualConsumerPubKey)
+	actualProviderConsAddr, found := providerKeeper.GetValidatorByConsumerAddr(ctx, "chainID", types.NewConsumerConsAddress(consumerAddr))
```

Comment:
we should definitely test that it is possible to assign a key the "normal way" after doing it like this. not sure if here is the right place to do it, or if it is easier with higher-level tests, e.g. e2e

Reply:
Conceptually, if we exclude opting in, a `MsgOptIn` (with a consumer key) followed by a `MsgAssign` message is the same as two `MsgAssign`s the one after the other. If we can call twice or more `MsgAssign` today, I don't see why `MsgOptIn` would cause issues with this. 
In any case, we could probably test this in an e2e test.